File: packages/server/src/contracts.ts
export type Awaitable<T> = T | Promise<T>;
export interface RequestContext {
  client?: string;
  query?: Record<string, string>;
}
export interface AuthRequest extends RequestContext {
  sessionId: string;
  headers: Record<string, string>;
  targetDir: string;
  meta: unknown;
}
export interface AuthResult {
  baseDir?: string;
  targetDir?: string;
  maxFiles?: number;
  maxTotalBytes?: number;
  context?: Record<string, unknown>;
  /** Automatically attempt publication after successful processing. */
  autoPublish?: boolean;
  /** Allow the browser to request publication. Backend publish remains available. */
  clientPublish?: boolean;
}
export interface CommitEvent {
  sessionId: string;
  targetDir: string;
  baseDir: string;
  stagingDir: string;
  files: number;
  bytes: number;
  context: Record<string, unknown>;
  meta: unknown;
}
export interface FileMapRequest {
  sessionId: string;
  path: string;
  name: string;
  size: number;
  targetDir: string;
  context: Record<string, unknown>;
  meta: unknown;
}
export interface StagedFile {
  path: string;
  size: number;
  mtime: number;
  localPath: string;
}
export interface Limits {
  concurrency: number;
  maxParts: number;
  batchBytes: number;
  partBytes: number;
}
export interface Options {
  baseDir: string;
  limits?: Partial<Limits>;
  ttlMs?: number;
  sweepIntervalMs?: number;
  maxMetaBytes?: number;
  maxContextBytes?: number;
  prefix?: string;
  autoPublish?: boolean;
  clientPublish?: boolean;
  authorize: (request: AuthRequest) => Awaitable<AuthResult | null>;
  mapFile?: (file: FileMapRequest) => Awaitable<string | null>;
  /** true publishes, false keeps staged; undefined uses the session autoPublish setting. */
  onCommitted?: (event: CommitEvent) => Awaitable<boolean | void>;
  onError?: (event: {
    hook: string;
    sessionId: string;
    error: unknown;
  }) => void;
}

File: server/mfup-core/mfup_core/contracts.py
from __future__ import annotations

from typing import Any, Awaitable, Callable, TypedDict


class AuthRequest(TypedDict):
    sessionId: str
    headers: dict[str, str]
    client: str
    query: dict[str, str]
    targetDir: str
    meta: Any


class AuthResult(TypedDict, total=False):
    baseDir: str
    targetDir: str
    maxFiles: int
    maxTotalBytes: int
    context: dict[str, Any]
    autoPublish: bool
    clientPublish: bool


class CommitEvent(TypedDict):
    sessionId: str
    targetDir: str
    baseDir: str
    stagingDir: str
    files: int
    bytes: int
    context: dict[str, Any]
    meta: Any


class FileMapRequest(TypedDict):
    sessionId: str
    path: str
    name: str
    size: int
    targetDir: str
    context: dict[str, Any]
    meta: Any


class StagedFile(TypedDict):
    path: str
    size: int
    mtime: int
    localPath: str


Authorize = Callable[[AuthRequest], AuthResult | None | Awaitable[AuthResult | None]]
MapFile = Callable[[FileMapRequest], str | None | Awaitable[str | None]]
OnCommitted = Callable[[CommitEvent], bool | None | Awaitable[bool | None]]
